src/components/lib/crt/crt.c

```diff
@@ -288,23 +287,80 @@ crt_ns_vas_split(struct crt_ns_vas *new, struct crt_ns_vas *existing, struct crt
 	return 0;
 }
 
+static struct crt_ns_vas *
+ns_vas_parent(struct crt_ns_vas *vas)
+{
+	assert(vas);
+
+	while(vas->parent != 0) vas = vas->parent;
+	return vas;
+}
+
+int
+crt_ns_vas_shared(struct crt_comp *client, struct crt_comp *server)
+{
+	if (!client->ns_vas || !server->ns_vas) return 0;
+
+	struct crt_ns_vas *vas = client->ns_vas;
+
+	while (vas) {
+		if (vas == server->ns_vas) return 1;
+		vas = vas->parent;
+	}
+
+	return 0;
+}
+
+int
+crt_ulk_init(void)
+{
+	cos_ulk_info_init(cos_compinfo_get(cos_defcompinfo_curr_get()));
+	cos_ulk_map_in(cos_compinfo_get(cos_defcompinfo_curr_get())->pgtbl_cap);
+
+	return 0;
+}
+
+int
+crt_ulk_map_in(struct crt_comp *comp)
+{
+	assert(comp);
+
+	struct cos_compinfo *ci = cos_compinfo_get(comp->comp_res);
+	return cos_ulk_map_in(ci->pgtbl_cap);
+
+}
+
+/* Since the NS implementation is architecture-aware, it can provide this abstraction */
```

Comment:
Lineas's entire address space namespace implementation is architecture specific. I never liked that it was in `crt`. I could definitly take some time to move all the funcitonality to a architecture specific library and create some further abstractions in crt.

Reply:
> I'm not sure I understand this. This is code that will need to compile and run on ARM. Not that this specific logic would cause any issues with that, but this comment makes me question if it will run/compile on other architectures.
> 
> This is in the architecture independent part of the codebase, so we want it to compile and run on other architectures even if it, obviously, won't use the custom HW support.

Linnea's entire address space namespace implementation is architecture specific. I never liked that it was in crt. I could definitly take some time to move all the funcitonality to a architecture specific library and create some further abstractions in crt.